internal/packages/internal/packagerender/celctx/cel.go

```diff
@@ -0,0 +1,142 @@
+	"package-operator.run/internal/packages/internal/packagetypes"
+)
+
+var errInvalidReturnType = errors.New("invalid return type")
+
+// CelCtx contains a cel environment that is prepared with a tmplCtx and pre-evaluated macros.
+type CelCtx struct {
+	env    *cel.Env
+	ctxMap map[string]any
+}
+
+// New pre-evaluates the given macros against tmplCtx and exposes both tmplCtx + macro results to cel programs.
+func New(macros []manifests.PackageManifestCelMacro, tmplCtx *packagetypes.PackageRenderContext) (CelCtx, error) {
+	ctxMap, opts, err := unpackContext(tmplCtx)
+	if err != nil {
+		return CelCtx{}, fmt.Errorf("context unpacking error: %w", err)
+	}
+
+	// create CEL environment with context
+	env, err := cel.NewEnv(opts...)
+	if err != nil {
+		return CelCtx{}, fmt.Errorf("env error: %w", err)
+	}
+
+	cc := CelCtx{
+		env:    env,
+		ctxMap: ctxMap,
+	}
+
+	for _, m := range macros {
+		// make sure macro name is allowed
+		if !macroNameRegexp.MatchString(m.Name) {
+			return CelCtx{}, fmt.Errorf("%w: '%s'", ErrInvalidCELMacroName, m.Name)
+		}
+
+		// make sure name is unique and does not shadow a key in ctxMap
+		if _, ok := ctxMap[m.Name]; ok {
+			return CelCtx{}, fmt.Errorf("%w: '%s'", ErrDuplicateCELMacroName, m.Name)
+		}
+
+		result, err := cc.Evaluate(m.Expression)
+		if err != nil {
+			return CelCtx{}, fmt.Errorf("%w: '%s': %w", ErrCELMacroEvaluation, m.Name, err)
+		}
+
+		// store evaluation result in context
+		ctxMap[m.Name] = result
+
+		// store macro variable name declaration
+		opts = append(opts, cel.Variable(m.Name, cel.BoolType))
+	}
+
+	// recreate CEL environment with macro declarations
+	env, err = cel.NewEnv(opts...)
+	if err != nil {
+		return CelCtx{}, fmt.Errorf("env error: %w", err)
+	}
+	cc.env = env
+
+	return cc, nil
+}
+
+// Evaluate CEL expressions against the prepared template context and macro results.
+func (cc *CelCtx) Evaluate(expr string) (bool, error) {
+	// compile CEL expression
+	ast, issues := cc.env.Compile(expr)
+	if issues != nil && issues.Err() != nil {
+		return false, fmt.Errorf("compile error: %w", issues.Err())
+	}
+
+	// create program
+	program, err := cc.env.Program(ast)
+	if err != nil {
+		return false, fmt.Errorf("program construction error: %w", err)
+	}
+
+	// evaluate the expression with context input
+	out, _, err := program.Eval(cc.ctxMap)
+	if err != nil {
+		return false, fmt.Errorf("evaluation error: %w", err)
+	}
+
+	// make sure that result type is 'bool'
+	if !reflect.DeepEqual(out.Type(), cel.BoolType) {
+		return false, newErrInvalidReturnType(ast.OutputType(), cel.BoolType)
+	}
+
+	return out.Value().(bool), nil
+}
+
+func newErrInvalidReturnType(actual, expected *cel.Type) error {
+	return fmt.Errorf("%w: %v, expected %v", errInvalidReturnType, actual, expected)
+}
+
+func unpackContext(tmplCtx *packagetypes.PackageRenderContext) (map[string]any, []cel.EnvOption, error) {
+	// TODO ask why?
+	if tmplCtx == nil {
+		return map[string]any{}, []cel.EnvOption{}, nil
+	}
```

Comment:
@pbabic-redhat Do you know why and how tmplCtx is `supposed` to be nil here?

Reply:
ATM, this is only used in tests. However, since the function receives a pointer, I thought it might be safer to check it. That way, it's less susceptible to breaking in the future. But I can totally remove it and modify the tests. WDYT?

Reply:
Ah understood!

One could argue that swallowing an error state (tmplCtx being a nil-pointer) and covering it up by returning an empty map is surprising and "broken" behavior if we don't expect `tmplCtx` to be ever actually be a nil-pointer.

I'm not so sure if it's worth removing it - use your best judgement. :)

PS: while we're already chatting here - I've talked to @thetechnick about the PR and we would like to request some API changes. We think it makes sense to structure and name the api (combined with the folder conditions PR) like this:

```
spec:
  conditionals:
    snippets:
    - name: xxx
      expression: yyy != zzz
    paths:
    - glob: 'foo/bar/**'
      expression: snippet.xxx && answer == 42 # use the snippet
```

effectively:
- moving the api under an optional key `.spec.conditionals`
- changing the name `macros` to `snippets`
- namespacing the snippet result values into `snippets` in the cel env

an alternative naming convention could look like this:

```
spec:
  conditionalFiltering:
    namedConditions:
    - name: xxx
      expressoin: yyy != zzz
    globs:
    - glob: 'foo/bar/**'
      expression: cond.xxx && answer == 42 # use the named condition
```

Do you think that makes sense? Pls let me know if you want help implementing the changes.

Reply:
Thanks for the feedback @erdii! Sorry I took so long to respond. I really like the idea of grouping the features into a single field. Together with the renaming in the first sample, I think it makes for a more intuitive UX.

I have a question regarding the snippet namespacing. What is the motivation behind it? I understand that it would force the users to be more explicit, but it would also require them to write more boilerplate. So far, there are 4 template context variables injected into the environment. Do you think these are likely to clash with desirable snippet names?

Reply:
Thanks a lot! :) And you do not have to exuse yourself for taking some time. You're great at this and your pacing is too! :)

### Namespacing

Cons:
- More boilerplate code for users because they have to prefix `namespace.` to their condition names.

Pros:
- Backwards compatibility: We can easily add more names without breaking backwards compatiblity with packages that use the same name for a condition.
- Less surprising: A user won't have to question if a variable is an expression they can control (`namespace.foo`) or something that PKO injects for them.

The con can be countered by using a short namespace name. The backwards compatibility pro is a must have.

---

Now that the important stuff has been written - let's do some bikeshedding:

### Naming things

I'd prefer this naming convention
```
spec:
  conditionalFiltering:
    namedConditions:
    - name: xxx
      expression: yyy != zzz
    paths:
    - glob: 'foo/bar/**'
      expression: cond.xxx && answer == 42
```

over the first one because:
- we are using conditional expressions to filter a file tree => "Conditional Filtering" is a very unsurprising name for this
- we are optionally giving conditions a reusable name and store them in a list -> namedCondtions
- the conditions are stored and injected in a map `cond` which is short and concise.

Whereas in the first convention
- the key `conditionals` tells us what we are specifying but not what it does.
- the key `snippets` tells us the right thing (reusable code pieces) but it rubs me the wrong way because this code is only every evaluated once and then stuffed in a variable of the same name, so is it really reusable? kind of yes, kind of no. :shrug: